snowfall/tools/ali.py

```diff
@@ -196,3 +203,107 @@ def visualize(input: str,
     if ret != 0:
         raise Exception(f'Failed to run\n{cmd}\n'
                         f'The praat script content is:\n{command}')
+
+
+def compute_ali(lang_dir: str,
+                posts: str,
+                output_dir: str,
+                device: torch.device,
+                max_duration: int = 200,
+                output_beam_size: float = 8.0):
+    # First, load lang
+    lang_dir = Path(lang_dir)
+
+    symbol_table = k2.SymbolTable.from_file(lang_dir / 'words.txt')
+    phone_symbol_table = k2.SymbolTable.from_file(lang_dir / 'phones.txt')
+    phone_ids = get_phone_symbols(phone_symbol_table)
+
+    phone_ids_with_blank = [0] + phone_ids
+    ctc_topo = k2.arc_sort(build_ctc_topo(phone_ids_with_blank))
+
+    if not (lang_dir / 'HLG.pt').exists():
+        print('Loading L_disambig.fst.txt')
+        with open(lang_dir / 'L_disambig.fst.txt') as f:
+            L = k2.Fsa.from_openfst(f.read(), acceptor=False)
+        print('Loading G.fst.txt')
+        with open(lang_dir / 'G.fst.txt') as f:
+            G = k2.Fsa.from_openfst(f.read(), acceptor=False)
+        first_phone_disambig_id = find_first_disambig_symbol(
+            phone_symbol_table)
+        first_word_disambig_id = find_first_disambig_symbol(symbol_table)
+        HLG = compile_HLG(L=L,
+                          G=G,
+                          H=ctc_topo,
+                          labels_disambig_id_start=first_phone_disambig_id,
+                          aux_labels_disambig_id_start=first_word_disambig_id)
+        torch.save(HLG.as_dict(), lang_dir / 'HLG.pt')
+    else:
+        print('Loading pre-compiled HLG')
+        d = torch.load(lang_dir / 'HLG.pt')
+        HLG = k2.Fsa.from_dict(d)
+
+    HLG = HLG.to(device)
+    HLG.aux_labels = k2.ragged.remove_values_eq(HLG.aux_labels, 0)
```

Comment:
what is this line doing? It looks like it's "sparsifying" the aux_labels (word ids) but how does HLG known which labels correspond to which aux_labels after that?